// @airbnb/lunar/lib/components/List/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _Item = _interopRequireDefault(require("./Item"));

exports.Item = _Item.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class List extends _react.default.Component {
  render() {
    const {
      children,
      cx,
      horizontal,
      ordered,
      styles,
      wrap
    } = this.props;
    const Tag = ordered ? 'ol' : 'ul';
    return _react.default.createElement(Tag, {
      className: cx(styles.list, horizontal && styles.list_horizontal, horizontal && wrap && styles.list_horizontal_wrap)
    }, _react.default.Children.map(children, child => {
      if (!child) {
        return null;
      }

      if (horizontal) {
        return _react.default.cloneElement(child, {
          horizontal
        });
      }

      return child;
    }));
  }

}

_defineProperty(List, "propTypes", {
  children: _propTypes.default.any.isRequired,
  horizontal: _propTypes.default.bool,
  ordered: _propTypes.default.bool,
  wrap: _propTypes.default.bool
});

var _default = (0, _withStyles.default)(() => ({
  list: {
    listStyle: 'none',
    margin: 0,
    padding: 0
  },
  list_horizontal: {
    display: 'flex'
  },
  list_horizontal_wrap: {
    flexWrap: 'wrap'
  }
}))(List);

exports.default = _default;
