// @airbnb/lunar/lib/components/List/Item.js
"use strict";

exports.__esModule = true;
exports.default = exports.ListItem = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const paddingPropType = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('compact', 'spacious');

class ListItem extends _react.default.Component {
  render() {
    const {
      bordered,
      children,
      compact,
      cx,
      horizontal,
      spacious,
      styles
    } = this.props;
    return _react.default.createElement("li", {
      className: cx(!horizontal && bordered && styles.item_bordered, horizontal && bordered && styles.item_bordered_horizontal, !horizontal && compact && styles.item_compact, horizontal && compact && styles.item_compact_horizontal, !horizontal && spacious && styles.item_spacious, horizontal && spacious && styles.item_spacious_horizontal)
    }, children);
  }

}

exports.ListItem = ListItem;

_defineProperty(ListItem, "propTypes", {
  horizontal: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  bordered: _propTypes.default.bool,
  compact: paddingPropType,
  spacious: paddingPropType
});

_defineProperty(ListItem, "defaultProps", {
  bordered: false,
  compact: false,
  horizontal: false,
  spacious: false
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    ui,
    unit
  } = _ref;
  return {
    item_bordered: {
      borderTop: ui.border,
      ':last-child': {
        borderBottom: ui.border
      }
    },
    item_bordered_horizontal: {
      borderLeft: ui.border,
      ':last-child': {
        borderRight: ui.border
      }
    },
    item_compact: {
      paddingBottom: unit * 1.5,
      paddingTop: unit * 1.5
    },
    item_compact_horizontal: {
      paddingLeft: unit * 1.5,
      paddingRight: unit * 1.5
    },
    item_spacious: {
      paddingBottom: unit * 3,
      paddingTop: unit * 3
    },
    item_spacious_horizontal: {
      paddingLeft: unit * 3,
      paddingRight: unit * 3
    }
  };
})(ListItem);

exports.default = _default;
