// @airbnb/lunar/esm/components/DataTable/columns/renderDataColumns.js
import React from 'react';
import { Column } from 'react-virtualized';
import DefaultRenderer from '../DefaultRenderer';
import Spacing from '../../Spacing';
import { DEFAULT_WIDTH_PROPERTIES } from '../constants';
export default function renderDataColumns(keys, editMode, onEdit, _ref) {
  let {
    columnMetadata,
    showColumnDividers,
    cx,
    styles,
    renderers,
    zebra,
    theme,
    selectable,
    expandable
  } = _ref;

  const renderCell = (key, isLeftmost) => row => {
    const {
      metadata
    } = row.rowData;
    const {
      isChild
    } = metadata;
    const customRenderer = renderers && renderers[key];
    const indentSize = !expandable || !isLeftmost ? 2 : 2.5;
    const spacing = isChild || !((expandable || selectable) && isLeftmost) ? indentSize : 0;
    const rendererArguments = {
      row,
      keyName: key,
      editMode,
      onEdit,
      zebra: zebra || false,
      theme
    };

    if (metadata && metadata.colSpanKey && renderers) {
      if (isLeftmost) {
        const colSpanRenderer = renderers[metadata.colSpanKey];

        if (colSpanRenderer) {
          return React.createElement(colSpanRenderer, rendererArguments);
        }
      }
    }

    const contents = React.createElement(customRenderer || DefaultRenderer, rendererArguments);
    return React.createElement("div", {
      className: cx(styles && styles.row)
    }, React.createElement("div", {
      className: cx(styles && styles.row_inner)
    }, React.createElement(Spacing, {
      left: spacing,
      right: 2
    }, contents || '')));
  };

  return keys.map((key, idx) => {
    const widthPropertiesOptions = ['maxWidth', 'minWidth', 'width', 'flexGrow', 'flexShrink'];
    const widthProperties = {};
    widthPropertiesOptions.forEach(property => {
      widthProperties[property] = columnMetadata && columnMetadata[key] !== undefined && columnMetadata[key][property] !== undefined ? columnMetadata[key][property] : DEFAULT_WIDTH_PROPERTIES[property];
    });
    const isLeftmost = idx === 0;
    const isRightmost = idx === keys.length - 1;
    return React.createElement(Column, {
      dataKey: key,
      key: key,
      label: key,
      width: widthProperties.width,
      flexGrow: widthProperties.flexGrow,
      flexShrink: widthProperties.flexShrink,
      maxWidth: widthProperties.maxWidth,
      minWidth: widthProperties.minWidth,
      cellRenderer: renderCell(key, isLeftmost),
      className: cx(styles && styles.column, showColumnDividers && !isRightmost && styles && styles.column_divider)
    });
  });
}
