// @airbnb/lunar/lib/components/Divider/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Divider = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _Spacing = _interopRequireDefault(require("../Spacing"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A horizontal divider. */
class Divider extends _react.default.Component {
  render() {
    const {
      cx,
      styles,
      bottom,
      short,
      top
    } = this.props;
    return _react.default.createElement(_Spacing.default, {
      bottom: bottom,
      top: top
    }, _react.default.createElement("div", {
      className: cx(styles.divider, short && styles.divider_short)
    }));
  }

}

exports.Divider = Divider;

_defineProperty(Divider, "propTypes", {
  bottom: _propTypes.default.any,
  short: _propTypes.default.bool,
  top: _propTypes.default.any
});

_defineProperty(Divider, "defaultProps", {
  bottom: 2,
  short: false,
  top: 2
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    ui,
    unit
  } = _ref;
  return {
    divider: {
      borderBottom: ui.border
    },
    divider_short: {
      width: 4 * unit
    }
  };
})(Divider);

exports.default = _default;
