// interweave-autolink/esm/Link.js
import _pt from "prop-types";

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';

var Link = function (_React$PureComponent) {
  _inheritsLoose(Link, _React$PureComponent);

  function Link() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Link.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        href = _this$props.href,
        onClick = _this$props.onClick,
        newWindow = _this$props.newWindow;
    return React.createElement("a", {
      href: href,
      target: newWindow ? '_blank' : undefined,
      onClick: onClick,
      rel: "noopener noreferrer"
    }, children);
  };

  return Link;
}(React.PureComponent);

_defineProperty(Link, "propTypes", {
  children: _pt.node.isRequired,
  href: _pt.string.isRequired,
  key: _pt.oneOfType([_pt.string, _pt.number]),
  newWindow: _pt.bool,
  onClick: _pt.func
});

_defineProperty(Link, "defaultProps", {
  newWindow: false,
  onClick: null
});

export { Link as default };
