// interweave-autolink/esm/Url.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import Link from './Link';

var Url = function (_React$PureComponent) {
  _inheritsLoose(Url, _React$PureComponent);

  function Url() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Url.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        urlParts = _this$props.urlParts,
        props = _objectWithoutPropertiesLoose(_this$props, ["children", "urlParts"]);

    var url = children;

    if (!url.match(/^https?:\/\//)) {
      url = "http://" + url;
    }

    return React.createElement(Link, _extends({}, props, {
      href: url
    }), children);
  };

  return Url;
}(React.PureComponent);

_defineProperty(Url, "propTypes", {
  children: _pt.string.isRequired,
  urlParts: _pt.shape({
    auth: _pt.string.isRequired,
    fragment: _pt.string.isRequired,
    host: _pt.string.isRequired,
    path: _pt.string.isRequired,
    port: _pt.oneOfType([_pt.string, _pt.number]).isRequired,
    query: _pt.string.isRequired,
    scheme: _pt.string.isRequired
  }).isRequired
});

export { Url as default };
