// @airbnb/lunar/lib/components/Loader/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Loader = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A small 3-dot loading indicator. */
class Loader extends _react.default.Component {
  render() {
    const {
      cx,
      styles,
      inline,
      inverted,
      large,
      static: isStatic
    } = this.props;
    return _react.default.createElement("div", {
      className: cx(styles.loader, inline && styles.loader_inline, !isStatic && !inline && styles.loader_absolute)
    }, [1, 2, 3].map(no => _react.default.createElement("span", {
      key: no,
      className: cx(styles.dot, styles["dot_" + no], large && styles.dot_large, inverted && styles.dot_inverted)
    })));
  }

}

exports.Loader = Loader;

_defineProperty(Loader, "propTypes", {
  inline: _propTypes.default.bool,
  inverted: _propTypes.default.bool,
  large: _propTypes.default.bool,
  static: _propTypes.default.bool
});

_defineProperty(Loader, "defaultProps", {
  inline: false,
  inverted: false,
  large: false,
  static: false
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    unit
  } = _ref;
  return {
    loader: {
      margin: '0 auto',
      textAlign: 'center'
    },
    loader_absolute: {
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: 'translateX(-50%) translateY(-50%)'
    },
    loader_inline: {
      display: 'inline-block'
    },
    dot: {
      width: 6,
      height: 6,
      marginRight: unit / 2,
      borderRadius: '100%',
      display: 'inline-block',
      animationName: {
        name: 'fade',
        '0%, 80%, 100%': {
          opacity: 0
        },
        '30%, 50%': {
          opacity: 1
        }
      },
      animationDuration: '0.8s',
      animationIterationCount: 'infinite',
      animationTimingFunction: 'linear',
      animationFillMode: 'both',
      verticalAlign: 'middle',
      backgroundColor: color.core.primary[3]
    },
    dot_inverted: {
      backgroundColor: color.accent.bg
    },
    dot_large: {
      width: 12,
      height: 12,
      marginRight: unit
    },
    dot_1: {
      animationDelay: '-0.3s'
    },
    dot_2: {
      animationDelay: '-0.15s'
    }
  };
})(Loader);

exports.default = _default;
