// @airbnb/lunar/lib/components/LabeledDivider/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.LabeledDivider = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Text = _interopRequireDefault(require("../Text"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A horizontal divider with a label. */
class LabeledDivider extends _react.default.Component {
  render() {
    const {
      cx,
      styles,
      label
    } = this.props;
    return _react.default.createElement("div", {
      className: cx(styles.rule)
    }, _react.default.createElement(_Text.default, {
      small: true,
      bold: true,
      inline: true
    }, label));
  }

}

exports.LabeledDivider = LabeledDivider;

_defineProperty(LabeledDivider, "propTypes", {
  label: _propTypes.default.any.isRequired
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    ui
  } = _ref;
  return {
    rule: {
      borderBottom: ui.border
    }
  };
})(LabeledDivider);

exports.default = _default;
