// @airbnb/lunar/lib/components/Table/Cell.js
"use strict";

exports.__esModule = true;
exports.default = exports.TableCell = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** An individual table cell. */
class TableCell extends _react.default.Component {
  render() {
    const _this$props = this.props,
          {
      cx,
      centerAlign,
      children,
      header,
      startAlign,
      endAlign,
      styles,
      truncate,
      wrap
    } = _this$props,
          props = _objectWithoutPropertiesLoose(_this$props, ["cx", "centerAlign", "children", "header", "startAlign", "endAlign", "styles", "truncate", "wrap"]);

    const Tag = header ? 'th' : 'td';
    return _react.default.createElement(Tag, _extends({}, props, {
      className: cx(truncate && styles.cell_truncate, startAlign && styles.cell_left, centerAlign && styles.cell_center, endAlign && styles.cell_right, wrap && styles.cell_wrap)
    }), children);
  }

}

exports.TableCell = TableCell;

_defineProperty(TableCell, "propTypes", {
  centerAlign: _propTypes.default.bool,
  children: _propTypes.default.node,
  colSpan: _propTypes.default.number,
  endAlign: _propTypes.default.bool,
  header: _propTypes.default.bool,
  startAlign: _propTypes.default.bool,
  truncate: _propTypes.default.bool,
  wrap: _propTypes.default.bool
});

_defineProperty(TableCell, "defaultProps", {
  centerAlign: false,
  endAlign: false,
  header: false,
  startAlign: false,
  truncate: false,
  wrap: false
});

var _default = (0, _withStyles.default)(() => ({
  cell_truncate: {
    overflow: 'hidden',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    ':hover': {
      maxWidth: 'none',
      overflow: 'inherit',
      whiteSpace: 'inherit',
      wordWrap: 'break-word'
    }
  },
  cell_left: {
    textAlign: 'left'
  },
  cell_center: {
    textAlign: 'center'
  },
  cell_right: {
    textAlign: 'right'
  },
  cell_wrap: {
    whiteSpace: 'normal',
    wordWrap: 'break-word'
  }
}))(TableCell);

exports.default = _default;
