// @airbnb/lunar/esm/components/Divider/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';
import Spacing from '../Spacing';

/** A horizontal divider. */
export class Divider extends React.Component {
  render() {
    const {
      cx,
      styles,
      bottom,
      short,
      top
    } = this.props;
    return React.createElement(Spacing, {
      bottom: bottom,
      top: top
    }, React.createElement("div", {
      className: cx(styles.divider, short && styles.divider_short)
    }));
  }

}

_defineProperty(Divider, "propTypes", {
  bottom: _pt.any,
  short: _pt.bool,
  top: _pt.any
});

_defineProperty(Divider, "defaultProps", {
  bottom: 2,
  short: false,
  top: 2
});

export default withStyles((_ref) => {
  let {
    ui,
    unit
  } = _ref;
  return {
    divider: {
      borderBottom: ui.border
    },
    divider_short: {
      width: 4 * unit
    }
  };
})(Divider);
