// @airbnb/lunar/esm/components/Grid/Col.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';

/** A column within a grid. */
export class Col extends React.Component {
  render() {
    const {
      cx,
      children,
      offset,
      span,
      styles
    } = this.props;
    return React.createElement("div", {
      "data-span": span,
      "data-offset": offset,
      className: cx(styles.col, styles["span" + span], styles["offset" + offset])
    }, children);
  }

}

_defineProperty(Col, "propTypes", {
  children: _pt.any.isRequired,
  offset: _pt.oneOf([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]),
  span: _pt.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]).isRequired
});

_defineProperty(Col, "defaultProps", {
  offset: 0
});

export default withStyles((_ref) => {
  let {
    unit
  } = _ref;
  const spans = {};
  const offsets = {};
  Array.from({
    length: 12
  }, (v, k) => {
    const span = k + 1;
    const offset = k;
    const width = 100 / (12 / span);
    spans["span" + span] = {
      flexBasis: width + "%",
      maxWidth: width + "%"
    };
    offsets["offset" + offset] = {
      marginLeft: offset > 0 ? 100 / (12 / offset) + "%" : 0
    };
    return span;
  });
  return _extends({
    col: {
      flex: '0 0 auto',
      paddingLeft: unit,
      paddingRight: unit
    }
  }, spans, {}, offsets);
})(Col);
