// @airbnb/lunar/lib/components/TextArea/Proofreader/Mark.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Mark extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "ref", _react.default.createRef());

    _defineProperty(this, "handleSelect", () => {
      const {
        ref
      } = this;

      if (this.props.selected && ref.current) {
        this.props.onSelect(ref.current.offsetTop + ref.current.offsetHeight, ref.current.offsetLeft);
      }
    });
  }

  componentDidMount() {
    this.handleSelect();
  }

  componentDidUpdate(prevProps) {
    if (this.props.selected !== prevProps.selected) {
      this.handleSelect();
    }
  }

  render() {
    const {
      cx,
      children,
      selected,
      alwaysHighlight,
      styles
    } = this.props;
    const highlight = selected || alwaysHighlight;
    return _react.default.createElement("mark", {
      className: cx(styles.mark, highlight && styles.mark_highlight),
      ref: this.ref
    }, children);
  }

}

_defineProperty(Mark, "propTypes", {
  children: _propTypes.default.any.isRequired,
  selected: _propTypes.default.bool.isRequired,
  onSelect: _propTypes.default.func.isRequired,
  alwaysHighlight: _propTypes.default.bool
});

_defineProperty(Mark, "defaultProps", {
  alwaysHighlight: false
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    ui
  } = _ref;
  return {
    mark: {
      position: 'relative',
      color: 'transparent',
      backgroundColor: 'transparent',
      opacity: 0.75,
      padding: 1,
      margin: -1,
      transition: 'opacity .2s, background .2s',
      '::after': {
        position: 'absolute',
        display: 'block',
        content: '""',
        bottom: 0,
        left: 0,
        right: 0,
        height: 2,
        borderRadius: ui.borderRadius,
        backgroundColor: color.core.danger[2]
      }
    },
    mark_highlight: {
      opacity: ui.disabledOpacity,
      borderTopLeftRadius: ui.borderRadius,
      borderTopRightRadius: ui.borderRadius,
      backgroundColor: color.core.danger[2]
    }
  };
}, {
  extendable: true
})(Mark);

exports.default = _default;
