// @airbnb/lunar/lib/components/Interweave/factories/Email.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Link = _interopRequireDefault(require("../../Link"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Email extends _react.default.PureComponent {
  render() {
    const {
      children,
      large,
      newWindow,
      small
    } = this.props;
    return _react.default.createElement(_Link.default, {
      baseline: true,
      href: "mailto:" + children,
      openInNewWindow: newWindow,
      small: small,
      large: large
    }, children);
  }

}

exports.default = Email;

_defineProperty(Email, "propTypes", {
  small: _propTypes.default.bool,
  large: _propTypes.default.bool
});

_defineProperty(Email, "defaultProps", {
  large: false,
  newWindow: false,
  small: false
});
