// @airbnb/lunar/esm/components/Shimmer/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';

/** A loading indicator bar that shimmers. */
export class Shimmer extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "randomWidth", this.props.width === 'random' ? Math.round(Math.random() * (90 - 30) + 30) + "%" : '');
  }

  render() {
    const {
      cx,
      block,
      height,
      radius,
      styles,
      width
    } = this.props;
    return React.createElement("span", {
      "aria-busy": "true",
      style: {
        borderRadius: radius,
        width: this.randomWidth || width,
        height
      },
      className: cx(styles.shimmer, block && styles.shimmer_block)
    });
  }

}

_defineProperty(Shimmer, "propTypes", {
  block: _pt.bool,
  height: _pt.oneOfType([_pt.number, _pt.string]),
  width: _pt.oneOfType([_pt.number, _pt.string]),
  radius: _pt.string
});

_defineProperty(Shimmer, "defaultProps", {
  block: false,
  height: '1.25ex',
  radius: '0.125em',
  width: '60%'
});

export default withStyles(() => ({
  shimmer: {
    animationDirection: 'alternate',
    animationDuration: '1s',
    animationFillMode: 'forwards',
    animationIterationCount: 'infinite',
    animationName: {
      name: 'shimmer',
      from: {
        opacity: 0.1
      },
      to: {
        opacity: 0.3
      }
    },
    animationTimingFunction: 'ease-in-out',
    backgroundColor: 'currentColor',
    display: 'inline-block',
    position: 'relative',
    verticalAlign: 'middle'
  },
  shimmer_block: {
    display: 'block'
  }
}))(Shimmer);
