// @airbnb/lunar/lib/components/HierarchyPicker/Hierarchy/ItemDescription.js
"use strict";

exports.__esModule = true;
exports.default = exports.ItemDescription = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../../composers/withStyles"));

var _Text = _interopRequireDefault(require("../../Text"));

var _Translate = _interopRequireDefault(require("../../Translate"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ItemDescription extends _react.default.Component {
  render() {
    const {
      cx,
      item,
      styles
    } = this.props;
    return _react.default.createElement("div", null, _react.default.createElement(_Text.default, {
      bold: true
    }, _react.default.createElement(_Translate.default, {
      k: "lunar.common.description",
      phrase: "Description",
      context: "Description of item"
    })), _react.default.createElement("div", {
      className: cx(styles.description)
    }, _react.default.createElement(_Text.default, null, item.description)));
  }

}

exports.ItemDescription = ItemDescription;

_defineProperty(ItemDescription, "propTypes", {
  item: _propTypes.default.any.isRequired
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    unit
  } = _ref;
  return {
    description: {
      marginTop: unit / 2
    }
  };
})(ItemDescription);

exports.default = _default;
