// @airbnb/lunar/lib/components/HierarchyPicker/Search/Highlight.js
"use strict";

exports.__esModule = true;
exports.default = exports.Highlight = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Highlight extends _react.default.Component {
  render() {
    const {
      cx,
      fallback,
      match,
      styles,
      word: searchWord
    } = this.props;

    if (!match) {
      return _react.default.createElement(_react.default.Fragment, null, fallback);
    }

    const {
      value,
      indices = []
    } = match;
    const matchIndices = [...indices]; // clean ref

    const output = [];
    let pair = matchIndices.shift();
    let substr = '';

    for (let i = 0; i < value.length; i += 1) {
      // if substr exists and we reach the start index of a match, push current substr and reset
      if (pair && substr && i === pair[0]) {
        output.push(_react.default.createElement("span", {
          key: i + "-start"
        }, substr));
        substr = '';
      }

      substr += value.charAt(i); // push highlight when we reach the end index of a match

      if (pair && i === pair[1]) {
        if (substr) {
          output.push(_react.default.createElement("span", {
            key: "end-" + i,
            className: cx(styles.highlight, substr.trim().toLowerCase() === searchWord && styles.highlight_dark)
          }, _react.default.createElement("mark", null, substr)));
        }

        substr = '';
        pair = matchIndices.shift();
      }
    }

    if (substr) {
      output.push(_react.default.createElement("span", {
        key: "last"
      }, substr));
    }

    return _react.default.createElement(_react.default.Fragment, null, output);
  }

}

exports.Highlight = Highlight;

_defineProperty(Highlight, "propTypes", {
  fallback: _propTypes.default.string,
  match: _propTypes.default.oneOfType([_propTypes.default.any, _propTypes.default.oneOf([null])]),
  word: _propTypes.default.string
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    ui
  } = _ref;
  return {
    highlight: {
      borderRadius: ui.borderRadius,
      backgroundColor: color.core.warning[0],
      '@selectors': {
        '> mark': {
          backgroundColor: 'transparent',
          color: 'inherit',
          position: 'relative',
          whiteSpace: 'nowrap'
        }
      }
    },
    highlight_dark: {
      backgroundColor: color.core.warning[3]
    }
  };
})(Highlight);

exports.default = _default;
