// @airbnb/lunar/esm/components/Menu/Row.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';

/** A non-interactive row within a menu. */
export class MenuRow extends React.Component {
  render() {
    const {
      cx,
      children,
      spacious,
      styles
    } = this.props;
    return React.createElement("li", {
      role: "none"
    }, React.createElement("div", {
      className: cx(styles.item, spacious && styles.item_spacious)
    }, children));
  }

}

_defineProperty(MenuRow, "propTypes", {
  children: _pt.any.isRequired,
  spacious: _pt.bool
});

_defineProperty(MenuRow, "defaultProps", {
  spacious: false
});

export default withStyles((_ref) => {
  let {
    unit
  } = _ref;
  return {
    item: {
      display: 'block',
      width: '100%',
      padding: unit,
      border: 0,
      textAlign: 'left'
    },
    item_spacious: {
      padding: unit * 2
    }
  };
})(MenuRow);
