// @airbnb/lunar/esm/components/private/BaseRadioButton.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import IconRecord from '@airbnb/lunar-icons/lib/interface/IconRecord';
import IconRemove from '@airbnb/lunar-icons/lib/interface/IconRemove';
import withStyles from '../../composers/withStyles';
import FormInput from './FormInput';
import buildInputStyles from '../../themes/buildInputStyles';

class BaseRadioButton extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleChange", event => {
      this.props.onChange(event.currentTarget.checked, event.currentTarget.value, event);
    });

    _defineProperty(this, "renderRadioButton", () => {
      const _this$props = this.props,
            {
        cx,
        checked,
        disabled,
        hideLabel,
        id,
        invalid,
        indeterminate,
        styles
      } = _this$props,
            restProps = _objectWithoutPropertiesLoose(_this$props, ["cx", "button", "checked", "children", "disabled", "hideLabel", "id", "invalid", "indeterminate", "styles"]);

      return React.createElement("label", {
        htmlFor: id,
        className: cx(styles.radio, hideLabel && styles.radio_hideLabel)
      }, React.createElement(FormInput, _extends({}, restProps, {
        checked: checked,
        disabled: disabled,
        id: id,
        invalid: invalid,
        onChange: this.handleChange,
        tagName: "input",
        type: "radio",
        hidden: true,
        optional: true
      })), React.createElement("span", {
        className: cx(styles.input, indeterminate && styles.input_indeterminate, checked && styles.input_checked, invalid && styles.input_invalid, disabled && styles.input_disabled)
      }, checked && React.createElement("span", {
        className: cx(styles.bullet)
      }, React.createElement(IconRecord, {
        decorative: true,
        size: "1.35em"
      })), indeterminate && React.createElement("span", {
        className: cx(styles.indeterminate)
      }, React.createElement(IconRemove, {
        decorative: true,
        size: "1.65em"
      }))));
    });
  }

  render() {
    const {
      cx,
      button,
      checked,
      children,
      disabled,
      id,
      invalid,
      indeterminate,
      styles
    } = this.props;

    if (!button) {
      return this.renderRadioButton();
    }

    return React.createElement("label", {
      htmlFor: id,
      className: cx(styles.button, indeterminate && styles.input_indeterminate, checked && styles.button_checked, invalid && styles.button_invalid, disabled && styles.button_disabled)
    }, this.renderRadioButton(), React.createElement("div", {
      className: cx(styles.children)
    }, children));
  }

}

_defineProperty(BaseRadioButton, "propTypes", {
  button: _pt.bool,
  children: _pt.node,
  hideLabel: _pt.bool,
  onChange: _pt.func.isRequired,
  indeterminate: _pt.bool
});

_defineProperty(BaseRadioButton, "defaultProps", {
  checked: false,
  indeterminate: false
});

export default withStyles(theme => {
  const styles = buildInputStyles(theme);
  return _extends({}, styles, {
    radio: {
      padding: 0,
      margin: 0,
      width: 18,
      height: 18,
      display: 'block',
      position: 'relative',
      lineHeight: 0,
      // Add focus styling to the radio icons
      '@selectors': {
        '> input:focus + span': {
          boxShadow: "0 0 3px 3px " + theme.color.core.primary[2]
        }
      }
    },
    radio_hideLabel: {
      display: 'block',
      marginTop: 0
    },
    input: _extends({}, styles.input, {
      width: 18,
      height: 18,
      padding: 0,
      display: 'inline-block',
      cursor: 'pointer',
      borderRadius: '50%',
      position: 'relative',
      // Bullet
      textAlign: 'center',
      fontSize: 8,
      lineHeight: '16px',
      letterSpacing: 0
    }),
    bullet: {
      position: 'absolute',
      top: 1.5,
      left: 1.5
    },
    indeterminate: {
      position: 'absolute',
      top: 1,
      left: 0.5
    },
    children: {
      marginLeft: theme.unit
    }
  });
})(BaseRadioButton);
