// @airbnb/lunar/lib/components/FormErrorMessage/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.FormErrorMessage = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _StatusText = _interopRequireDefault(require("../StatusText"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Display a simple error message for use within forms. */
class FormErrorMessage extends _react.default.Component {
  render() {
    const {
      cx,
      id,
      error,
      styles
    } = this.props;

    if (!error) {
      return null;
    }

    return _react.default.createElement("div", {
      id: id + "-error",
      role: "alert",
      "aria-live": "polite",
      className: cx(styles.error)
    }, _react.default.createElement(_StatusText.default, {
      danger: true
    }, error));
  }

}

exports.FormErrorMessage = FormErrorMessage;

_defineProperty(FormErrorMessage, "propTypes", {
  id: _propTypes.default.string.isRequired,
  error: _propTypes.default.string
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    unit
  } = _ref;
  return {
    error: {
      marginTop: unit
    }
  };
})(FormErrorMessage);

exports.default = _default;
