// @airbnb/lunar/lib/components/ProfilePhotoGroup/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.ProfilePhotoGroup = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Horizontally align `ProfilePhoto`s in a stacked fashion. */
class ProfilePhotoGroup extends _react.default.Component {
  render() {
    const {
      cx,
      children,
      max,
      size,
      styles,
      theme
    } = this.props;
    const {
      unit
    } = theme;
    const margin = {
      marginLeft: -(size * (unit / 2.5))
    };

    let photos = _react.default.Children.toArray(children).filter(child => !!child);

    let remainder = 0;

    if (max && photos.length > max) {
      remainder = photos.length - max;
      photos = photos.slice(0, max);
    }

    return _react.default.createElement("div", {
      className: cx(styles.group)
    }, photos.map((photo, i) => _react.default.createElement("div", {
      key: i,
      className: cx(styles.cell, margin)
    }, _react.default.cloneElement(photo, {
      size,
      inline: false,
      square: false
    }))), remainder > 0 && _react.default.createElement("div", {
      key: "remainder",
      className: cx(styles.cell, margin)
    }, _react.default.createElement("div", {
      className: cx(styles.remainder, {
        height: size * unit,
        maxHeight: size * unit,
        maxWidth: size * unit,
        width: size * unit
      })
    }, _react.default.createElement("span", {
      className: cx(styles.remainderNumber)
    }, "+" + remainder))));
  }

}

exports.ProfilePhotoGroup = ProfilePhotoGroup;

_defineProperty(ProfilePhotoGroup, "propTypes", {
  children: _propTypes.default.any.isRequired,
  max: _propTypes.default.number,
  size: _propTypes.default.number
});

_defineProperty(ProfilePhotoGroup, "defaultProps", {
  max: 3,
  size: 5
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    font
  } = _ref;
  return {
    group: {
      display: 'flex',
      alignItems: 'stretch'
    },
    cell: {
      borderRadius: '50%',
      border: "1px solid " + color.base,
      background: color.core.neutral[6],
      ':first-of-type': {
        marginLeft: 0
      },
      ':empty': {
        display: 'none'
      }
    },
    remainder: _extends({}, font.textRegular, {
      position: 'relative'
    }),
    remainderNumber: {
      transform: 'translate(-50%, -50%)',
      position: 'absolute',
      top: '50%',
      left: '50%',
      color: color.base
    }
  };
}, {
  passThemeProp: true
})(ProfilePhotoGroup);

exports.default = _default;
