// @airbnb/lunar/lib/components/StatusText/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.StatusText = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _Text = _interopRequireDefault(require("../Text"));

var _constants = require("../../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const statusPropType = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)(..._constants.STATUSES);

/** Display a string of classified text with colorful statuses. */
class StatusText extends _react.default.Component {
  render() {
    const _this$props = this.props,
          {
      cx,
      children,
      danger,
      info,
      muted,
      notice,
      styles,
      success,
      warning
    } = _this$props,
          restProps = _objectWithoutPropertiesLoose(_this$props, ["cx", "children", "danger", "info", "muted", "notice", "styles", "success", "warning"]);

    return _react.default.createElement(_Text.default, restProps, _react.default.createElement("span", {
      className: cx(danger && styles.text_danger, info && styles.text_info, muted && styles.text_muted, notice && styles.text_notice, success && styles.text_success, warning && styles.text_warning)
    }, children));
  }

}

exports.StatusText = StatusText;

_defineProperty(StatusText, "propTypes", {
  children: _propTypes.default.any.isRequired,
  danger: statusPropType,
  info: statusPropType,
  muted: statusPropType,
  notice: statusPropType,
  success: statusPropType,
  warning: statusPropType
});

_defineProperty(StatusText, "defaultProps", {
  danger: false,
  info: false,
  muted: false,
  notice: false,
  success: false,
  warning: false
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color
  } = _ref;
  return {
    text_danger: {
      color: color.core.danger[4]
    },
    text_info: {
      color: color.core.primary[4]
    },
    text_muted: {
      color: color.muted
    },
    text_notice: {
      color: color.core.secondary[4]
    },
    text_success: {
      color: color.core.success[4]
    },
    text_warning: {
      color: color.core.warning[5]
    }
  };
})(StatusText);

exports.default = _default;
