// @airbnb/lunar/esm/components/SortCarets/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import IconCaretUp from '@airbnb/lunar-icons/lib/interface/IconCaretUp';
import IconCaretDown from '@airbnb/lunar-icons/lib/interface/IconCaretDown';
import withStyles from '../../composers/withStyles';

/** Carets to indicate sorting on a table. */
export class SortCarets extends React.Component {
  renderCaretUp() {
    const {
      cx,
      up,
      enableUp,
      styles
    } = this.props;
    return up && React.createElement("span", {
      className: cx(styles.caret, styles.caret_up, enableUp ? styles.caret_active : styles.caret_inactive)
    }, React.createElement(IconCaretUp, {
      decorative: true,
      size: "1.6em"
    }));
  }

  renderCaretDown() {
    const {
      cx,
      down,
      enableDown,
      styles
    } = this.props;
    return down && React.createElement("span", {
      className: cx(styles.caret, styles.caret_down, enableDown ? styles.caret_active : styles.caret_inactive)
    }, React.createElement(IconCaretDown, {
      decorative: true,
      size: "1.6em"
    }));
  }

  render() {
    const {
      cx,
      styles,
      up,
      down
    } = this.props;
    return React.createElement("span", {
      className: cx(styles.container, up && down && styles.container_full)
    }, this.renderCaretUp(), this.renderCaretDown());
  }

}

_defineProperty(SortCarets, "propTypes", {
  down: _pt.bool,
  enableDown: _pt.bool,
  enableUp: _pt.bool,
  up: _pt.bool
});

_defineProperty(SortCarets, "defaultProps", {
  down: false,
  enableDown: false,
  enableUp: false,
  up: false
});

export default withStyles((_ref) => {
  let {
    color,
    unit
  } = _ref;
  return {
    container: {
      display: 'inline-block',
      verticalAlign: 'middle',
      marginTop: -1,
      width: unit * 1.5,
      height: unit
    },
    container_full: {
      height: unit * 2
    },
    caret: {
      display: 'block',
      position: 'relative',
      width: unit * 1.5,
      height: unit,
      overflow: 'hidden'
    },
    caret_up: {
      '@selectors': {
        '> svg': {
          margin: '-.4em',
          marginTop: '-.5em'
        }
      }
    },
    caret_down: {
      '@selectors': {
        '> svg': {
          margin: '-.4em',
          marginTop: '-.6em'
        }
      }
    },
    caret_inactive: {
      color: color.core.neutral[3]
    },
    caret_active: {
      color: color.core.neutral[4]
    }
  };
})(SortCarets);
