// @airbnb/lunar/esm/components/Accordion/Item.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';
import ExpandableIcon from '../ExpandableIcon';

/**
 * A single accordion item. To be rendered amongst a collection of accordion items, within an accordion.
 */
export class AccordionItem extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleClick", () => {
      if (this.props.onClick) {
        this.props.onClick(this.props.index);
      }
    });
  }

  render() {
    const {
      cx,
      bordered,
      children,
      expanded,
      id,
      styles,
      theme,
      title
    } = this.props;
    return React.createElement("div", {
      className: cx(styles.item, bordered && styles.item_bordered)
    }, React.createElement("button", {
      className: cx(styles.title),
      "aria-controls": "accordion-body-" + id,
      "aria-selected": expanded,
      id: "accordion-title-" + id,
      onClick: this.handleClick,
      role: "tab",
      tabIndex: 0,
      type: "button"
    }, title && React.createElement("span", {
      className: cx(styles.titleText)
    }, title), React.createElement(ExpandableIcon, {
      expanded: !!expanded,
      size: theme.unit * 3
    })), React.createElement("section", {
      className: cx(styles.body, expanded && styles.body_expanded),
      "aria-hidden": !expanded,
      "aria-labelledby": "accordion-title-" + id,
      id: "accordion-body-" + id,
      role: "tabpanel"
    }, children));
  }

}

_defineProperty(AccordionItem, "propTypes", {
  bordered: _pt.bool,
  children: _pt.node,
  expanded: _pt.bool,
  id: _pt.string.isRequired,
  index: _pt.number,
  title: _pt.node,
  onClick: _pt.func
});

_defineProperty(AccordionItem, "defaultProps", {
  bordered: false,
  children: null,
  expanded: false
});

export default withStyles((_ref) => {
  let {
    color,
    pattern,
    ui,
    unit
  } = _ref;
  return {
    body: {
      display: 'none',
      padding: unit + "px " + unit * 2 + "px " + unit * 2 + "px"
    },
    body_expanded: {
      display: 'block'
    },
    item: {
      background: color.accent.bg
    },
    item_bordered: {
      borderTop: ui.border
    },
    title: _extends({}, pattern.resetButton, {
      display: 'flex',
      alignItems: 'center',
      padding: unit * 2,
      textAlign: 'left',
      width: '100%'
    }),
    titleText: {
      flex: '1'
    }
  };
}, {
  passThemeProp: true
})(AccordionItem);
