// interweave-autolink/lib/Hashtag.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Link = _interopRequireDefault(require("./Link"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Hashtag = function (_React$PureComponent) {
  _inheritsLoose(Hashtag, _React$PureComponent);

  function Hashtag() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Hashtag.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        encodeHashtag = _this$props.encodeHashtag,
        hashtagUrl = _this$props.hashtagUrl,
        preserveHash = _this$props.preserveHash,
        props = _objectWithoutPropertiesLoose(_this$props, ["children", "encodeHashtag", "hashtagUrl", "preserveHash"]);

    var hashtag = String(children);

    if (!preserveHash && hashtag.charAt(0) === '#') {
      hashtag = hashtag.slice(1);
    }

    if (encodeHashtag) {
      hashtag = encodeURIComponent(hashtag);
    }

    var url = hashtagUrl || '{{hashtag}}';

    if (typeof url === 'function') {
      url = url(hashtag);
    } else {
      url = url.replace('{{hashtag}}', hashtag);
    }

    return _react.default.createElement(_Link.default, _extends({}, props, {
      href: url
    }), children);
  };

  return Hashtag;
}(_react.default.PureComponent);

exports.default = Hashtag;

_defineProperty(Hashtag, "propTypes", {
  children: _propTypes.default.string.isRequired,
  encodeHashtag: _propTypes.default.bool,
  hashtagName: _propTypes.default.string.isRequired,
  hashtagUrl: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func]),
  preserveHash: _propTypes.default.bool
});

_defineProperty(Hashtag, "defaultProps", {
  encodeHashtag: false,
  hashtagUrl: '{{hashtag}}',
  preserveHash: false
});
