// interweave-autolink/lib/Link.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Link = function (_React$PureComponent) {
  _inheritsLoose(Link, _React$PureComponent);

  function Link() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Link.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        href = _this$props.href,
        onClick = _this$props.onClick,
        newWindow = _this$props.newWindow;
    return _react.default.createElement("a", {
      href: href,
      target: newWindow ? '_blank' : undefined,
      onClick: onClick,
      rel: "noopener noreferrer"
    }, children);
  };

  return Link;
}(_react.default.PureComponent);

exports.default = Link;

_defineProperty(Link, "propTypes", {
  children: _propTypes.default.node.isRequired,
  href: _propTypes.default.string.isRequired,
  key: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  newWindow: _propTypes.default.bool,
  onClick: _propTypes.default.func
});

_defineProperty(Link, "defaultProps", {
  newWindow: false,
  onClick: null
});
