// @airbnb/lunar/lib/components/ProgressBar/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.ProgressBar = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A bar to represent the progress to completion. */
class ProgressBar extends _react.default.Component {
  render() {
    const {
      cx,
      percent,
      leading,
      trailing,
      styles
    } = this.props;
    return _react.default.createElement("div", {
      className: cx(styles.wrapper)
    }, _react.default.createElement("div", {
      className: cx(styles.bar, !leading && styles.bar_leading, !trailing && styles.bar_trailing)
    }, _react.default.createElement("div", {
      className: cx(styles.bar, !leading && styles.bar_leading, !trailing && styles.bar_trailing, styles.progress),
      style: {
        width: percent + "%"
      }
    })));
  }

}

exports.ProgressBar = ProgressBar;

_defineProperty(ProgressBar, "propTypes", {
  trailing: _propTypes.default.bool,
  leading: _propTypes.default.bool,
  percent: (0, _airbnbPropTypes.between)({
    gte: 0,
    lte: 100
  }).isRequired
});

_defineProperty(ProgressBar, "defaultProps", {
  leading: false,
  trailing: false
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    unit,
    ui
  } = _ref;
  return {
    wrapper: {
      paddingTop: unit / 4,
      paddingBottom: unit / 4
    },
    bar: {
      height: unit / 2,
      background: color.core.primary[1]
    },
    bar_leading: {
      borderTopRightRadius: ui.borderRadius,
      borderBottomRightRadius: ui.borderRadius
    },
    bar_trailing: {
      borderTopLeftRadius: ui.borderRadius,
      borderBottomLeftRadius: ui.borderRadius
    },
    progress: {
      background: color.core.primary[6]
    }
  };
})(ProgressBar);

exports.default = _default;
