// @airbnb/lunar/lib/components/Button/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Button = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _ButtonOrLink = _interopRequireDefault(require("../private/ButtonOrLink"));

var _Loader = _interopRequireDefault(require("../Loader"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const sizingProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('small', 'large');

/** A standard button and or link for common UI interactions. */
class Button extends _react.default.Component {
  render() {
    const _this$props = this.props,
          {
      cx,
      block,
      borderless,
      children,
      disabled,
      invalid,
      inverted,
      large,
      loading,
      small,
      styles
    } = _this$props,
          restProps = _objectWithoutPropertiesLoose(_this$props, ["cx", "block", "borderless", "children", "disabled", "invalid", "inverted", "large", "loading", "small", "styles"]);

    return _react.default.createElement(_ButtonOrLink.default, _extends({}, restProps, {
      "aria-busy": loading,
      disabled: disabled,
      loading: loading,
      className: cx(styles.button, large && styles.button_large, small && styles.button_small, !large && !small && styles.button_regular, block && styles.button_block, (disabled || loading) && styles.button_disabled, (borderless || inverted) && styles.button_inverted, invalid && styles.button_invalid, borderless && styles.button_borderless, loading && styles.button_loading)
    }), loading ? _react.default.createElement(_Loader.default, {
      inline: true,
      inverted: !inverted
    }) : children);
  }

}

exports.Button = Button;

_defineProperty(Button, "propTypes", {
  inverted: _propTypes.default.bool,
  invalid: _propTypes.default.bool,
  borderless: _propTypes.default.bool,
  block: _propTypes.default.bool,
  large: sizingProp,
  small: sizingProp
});

_defineProperty(Button, "defaultProps", {
  block: false,
  borderless: false,
  disabled: false,
  invalid: false,
  inverted: false,
  large: false,
  loading: false,
  small: false
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    font,
    pattern,
    ui,
    unit,
    transition
  } = _ref;
  return {
    button: _extends({}, pattern.resetButton, {}, transition.box, {
      fontWeight: font.weights.bold,
      position: 'relative',
      color: color.base,
      backgroundColor: color.core.primary[3],
      border: ui.borderWidthThick + "px solid " + color.core.primary[3],
      borderRadius: ui.borderRadius,
      textAlign: 'center',
      '@selectors': {
        // Removes weird bonus padding from button in Firefox
        '::-moz-focus-inner': {
          border: 0,
          padding: 0,
          margin: 0
        },
        // Only show hover states on non-disabled
        ':not([disabled]):hover': {
          backgroundColor: color.core.primary[4],
          borderColor: color.core.primary[4]
        }
      }
    }),
    button_block: {
      display: 'block',
      width: '100%',
      whiteSpace: 'normal',
      overflow: 'hidden'
    },
    button_borderless: {
      borderColor: 'transparent',
      '@selectors': {
        ':not([disabled]):hover': {
          borderColor: color.accent.bgHover
        }
      }
    },
    button_disabled: _extends({}, pattern.disabled),
    button_invalid: {},
    button_inverted: {
      color: color.core.primary[3],
      backgroundColor: color.accent.bg,
      '@selectors': {
        ':not([disabled]):hover': {
          color: color.core.primary[4],
          backgroundColor: color.accent.bgHover
        }
      }
    },
    button_loading: {
      cursor: 'default'
    },
    button_small: _extends({}, pattern.smallButton, {
      minWidth: 6 * unit
    }),
    button_regular: _extends({}, pattern.regularButton, {
      minWidth: 8 * unit
    }),
    button_large: _extends({}, pattern.largeButton, {
      minWidth: 9 * unit
    })
  };
}, {
  extendable: true
})(Button);

exports.default = _default;
