// @airbnb/lunar/lib/components/Menu/Row.js
"use strict";

exports.__esModule = true;
exports.default = exports.MenuRow = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A non-interactive row within a menu. */
class MenuRow extends _react.default.Component {
  render() {
    const {
      cx,
      children,
      spacious,
      styles
    } = this.props;
    return _react.default.createElement("li", {
      role: "none"
    }, _react.default.createElement("div", {
      className: cx(styles.item, spacious && styles.item_spacious)
    }, children));
  }

}

exports.MenuRow = MenuRow;

_defineProperty(MenuRow, "propTypes", {
  children: _propTypes.default.any.isRequired,
  spacious: _propTypes.default.bool
});

_defineProperty(MenuRow, "defaultProps", {
  spacious: false
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    unit
  } = _ref;
  return {
    item: {
      display: 'block',
      width: '100%',
      padding: unit,
      border: 0,
      textAlign: 'left'
    },
    item_spacious: {
      padding: unit * 2
    }
  };
})(MenuRow);

exports.default = _default;
