// @airbnb/lunar/lib/components/Menu/Separator.js
"use strict";

exports.__esModule = true;
exports.default = exports.MenuSeparator = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A separator between menu items. */
class MenuSeparator extends _react.default.Component {
  render() {
    const {
      cx,
      styles
    } = this.props;
    return _react.default.createElement("li", {
      role: "separator"
    }, _react.default.createElement("hr", {
      className: cx(styles.separator)
    }));
  }

}

exports.MenuSeparator = MenuSeparator;

_defineProperty(MenuSeparator, "propTypes", {
  fakeProp: _propTypes.default.any
});

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    unit
  } = _ref;
  return {
    separator: {
      marginTop: unit,
      marginBottom: unit,
      padding: 0,
      border: 0,
      height: 1,
      background: color.accent.border
    }
  };
})(MenuSeparator);

exports.default = _default;
