// @airbnb/lunar/esm/components/NotchedBox/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';

function createPosition(offset) {
  if (typeof offset === 'number' && offset < 0) {
    return {
      right: Math.abs(offset)
    };
  }

  return {
    left: offset
  };
}

export const NOTCH_SIZE = 1.5;
export const NOTCH_SPACING = 1.5;

/** A container with a configurable arrow pointing outward. */
export class NotchedBox extends React.Component {
  render() {
    const {
      cx,
      styles,
      inverted,
      children,
      inline,
      notchBelow,
      notchOffset
    } = this.props;
    return React.createElement("div", {
      className: cx(styles.box, !inverted && styles.border, inline && styles.box_inline)
    }, React.createElement("div", {
      className: cx(styles.notch, notchOffset ? createPosition(notchOffset) : styles.notch_position, notchBelow && styles.notch_below, inverted ? styles.inverted : styles.border)
    }), React.createElement("div", {
      className: cx(styles.content, inverted && styles.inverted)
    }, children));
  }

}

_defineProperty(NotchedBox, "propTypes", {
  children: _pt.any.isRequired,
  inline: _pt.bool,
  inverted: _pt.bool,
  notchBelow: _pt.bool,
  notchOffset: _pt.oneOfType([_pt.number, _pt.string])
});

_defineProperty(NotchedBox, "defaultProps", {
  inline: false,
  inverted: false
});

export default withStyles((_ref) => {
  let {
    ui,
    color,
    unit
  } = _ref;
  const {
    border,
    borderRadius
  } = ui;
  const notchSide = NOTCH_SIZE * unit;
  const offset = -notchSide / 2;
  const left = unit + notchSide / Math.SQRT2;
  return {
    box: {
      position: 'relative',
      borderRadius
    },
    box_inline: {
      display: 'inline-block'
    },
    notch: {
      width: notchSide,
      height: notchSide,
      position: 'absolute',
      backgroundColor: color.core.neutral[0],
      transform: "translate(" + offset + "px, " + offset + "px) rotate(-45deg)"
    },
    notch_position: createPosition(left),
    notch_below: {
      bottom: -notchSide
    },
    border: {
      border
    },
    inverted: {
      borderColor: color.clear,
      backgroundColor: color.core.neutral[6]
    },
    content: {
      padding: NOTCH_SPACING * unit,
      position: 'relative',
      backgroundColor: color.core.neutral[0],
      borderRadius
    }
  };
})(NotchedBox);
