// @airbnb/lunar/lib/components/SteppedProgressBar/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.SteppedProgressBar = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _Step = _interopRequireDefault(require("./Step"));

exports.Step = _Step.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A progress bar separated into individual steps. */
class SteppedProgressBar extends _react.default.Component {
  render() {
    const {
      cx,
      children,
      styles
    } = this.props;

    const steps = _react.default.Children.count(children);

    return _react.default.createElement("div", {
      className: cx(styles.bar)
    }, _react.default.Children.map(children, (child, index) => _react.default.cloneElement(child, {
      first: index === 0,
      last: index === steps - 1
    })));
  }

}

exports.SteppedProgressBar = SteppedProgressBar;

_defineProperty(SteppedProgressBar, "propTypes", {
  children: (0, _airbnbPropTypes.childrenOfType)(_Step.default).isRequired
});

var _default = (0, _withStyles.default)(() => ({
  bar: {
    display: 'flex'
  }
}))(SteppedProgressBar);

exports.default = _default;
