// @airbnb/lunar/lib/components/SteppedProgressBar/Step.js
"use strict";

exports.__esModule = true;
exports.default = exports.Step = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _ProgressBar = _interopRequireDefault(require("../ProgressBar"));

var _Tooltip = _interopRequireDefault(require("../Tooltip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** An individual step within a stepped progress bar. */
class Step extends _react.default.Component {
  render() {
    const {
      cx,
      complete,
      first,
      last,
      label,
      styles
    } = this.props;

    const bar = _react.default.createElement(_ProgressBar.default, {
      leading: first || !first && !last,
      trailing: last || !first && !last,
      percent: complete ? 100 : 0
    });

    return _react.default.createElement("div", {
      className: cx(styles.step)
    }, label ? _react.default.createElement(_Tooltip.default, {
      content: label
    }, bar) : bar);
  }

}

exports.Step = Step;

_defineProperty(Step, "propTypes", {
  complete: _propTypes.default.bool,
  first: _propTypes.default.bool,
  label: _propTypes.default.node,
  last: _propTypes.default.bool
});

_defineProperty(Step, "defaultProps", {
  complete: false,
  first: false,
  label: null,
  last: false
});

var _default = (0, _withStyles.default)(() => ({
  step: {
    flexGrow: 1,
    marginRight: 2,
    ':last-child': {
      marginRight: 0
    },
    '@selectors': {
      // Fix tooltips mixed with flexbox
      '> *': {
        display: 'block',
        width: '100%'
      }
    }
  }
}))(Step);

exports.default = _default;
