// @airbnb/lunar/esm/components/HierarchyPicker/Hierarchy/ItemDescription.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../../composers/withStyles';
import Text from '../../Text';
import T from '../../Translate';
export class ItemDescription extends React.Component {
  render() {
    const {
      cx,
      item,
      styles
    } = this.props;
    return React.createElement("div", null, React.createElement(Text, {
      bold: true
    }, React.createElement(T, {
      k: "lunar.common.description",
      phrase: "Description",
      context: "Description of item"
    })), React.createElement("div", {
      className: cx(styles.description)
    }, React.createElement(Text, null, item.description)));
  }

}

_defineProperty(ItemDescription, "propTypes", {
  item: _pt.any.isRequired
});

export default withStyles((_ref) => {
  let {
    unit
  } = _ref;
  return {
    description: {
      marginTop: unit / 2
    }
  };
})(ItemDescription);
