// @airbnb/lunar/lib/components/Sheet/SheetArea.js
"use strict";

exports.__esModule = true;
exports.default = exports.SheetArea = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _SheetContext = _interopRequireDefault(require("./SheetContext"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Container in which to render sheets. */
class SheetArea extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      visible: false
    });

    _defineProperty(this, "setSheetVisible", visible => {
      this.setState({
        visible
      });
    });
  }

  render() {
    const {
      cx,
      styles,
      children
    } = this.props;
    const {
      visible
    } = this.state;
    return _react.default.createElement(_SheetContext.default.Provider, {
      value: this.setSheetVisible
    }, _react.default.createElement("div", {
      className: cx(visible && styles.sheet_visible)
    }, children));
  }

}

exports.SheetArea = SheetArea;

_defineProperty(SheetArea, "propTypes", {
  children: _propTypes.default.any.isRequired
});

var _default = (0, _withStyles.default)(() => ({
  sheet_visible: {
    position: 'relative',
    overflow: 'hidden'
  }
}))(SheetArea);

exports.default = _default;
